### PR TITLE
Full width result view

### LINKED ISSUE #512: Limit results box width
Right now, it doesn't appear that the result bubble size is suitably limited in the horizontal dimension. The limit is determined by the longest line of the file (assuming soft linewrap is disabled), not by the width of the containing window.

If the lines printed to stdout are long, the box expands to the point where the controls (close/copy/open in new tab) are no longer visible. This problem is especially noticeable if using multiple panes.

I would propose to limit the result box width to be smaller than the current pane width. Some extra padding on the right would be useful to not interrupt scrolling. This is because results boxes can interfere with scrolling through the file: partway through scrolling the mouse runs over a result box and starts scrolling the stdout contents instead. Having margins around the results box helps prevent this. (It's technically possible to avoid this by either using the minimap to scroll, or scrolling by placing the mouse over the line number indicators. But at least for me my muscle memory doesn't have me do that.)

In addition to providing a margin specified in terms of pixels, another option is to use the "preferred line length" setting for the current file and cut of the results box at that many characters.

Screenshot of the problem:
screen shot 2016-11-20 at 8 46 36 pm

### PR DESCRIPTION
**Pull request by lgeiger**
Here is my initial proposal for a more robust multiline output area. The "single line" and status outputs are left as is.

I'm not a web designer and I'm very open to suggestions and contributions on this design. In any way this PR should really simplify the output system which makes it possible to add support for themes in the future.
bildschirmfoto 2016-11-24 um 00 51 13

~~The one thing that doesn't work yet is scrolling inside the output cells. I would really appreciate your help on this.~~

Closes #512

### CODE REVIEW

File: lib/result-view.js [hunk 1]
```diff
@@ -54,6 +53,17 @@ export default class ResultView {
...
 
+    this.outputContainer.onmousewheel = (event) => {
+      const clientHeight = this.outputContainer.clientHeight;
+      const scrollHeight = this.outputContainer.scrollHeight;
+      const scrollTop = this.outputContainer.scrollTop;
+      const atTop = (scrollTop !== 0 && event.deltaY < 0);
+      const atBottom = (scrollTop !== scrollHeight - clientHeight && event.deltaY > 0);
+
+      if (clientHeight < scrollHeight && (atTop || atBottom)) {
+        event.stopPropagation();
```
Comment (lgeiger) [inline lib/result-view.js]: There's probably a better way of doing this, but this will work for now.

**Comment by n-riesco** [comment]
> The one thing that doesn't work yet is scrolling inside the output cells.

Wasn't this handled in CSS before? `style='overflow:auto'`? E.g.:

```html
<div style='overflow:auto;width:200;height:200;border:solid'>
<div style='width:300;height:300;background-color:pink'>300x300</div>
</div>

<div style='overflow:auto;width:200;height:200;border:solid'>
<div style='width:100;height:100;background-color:pink'>100x100</div>
</div>
```

**Comment by lgeiger** [comment]
Whoops I forgot to cross out my comment about the scrolling. I figured it out in the latest commit.

We used `outputContainer.onmousewheel = e => e.stopPropagation();` before to enable scrolling in the result bubble. But this blocks the scrolling of the text editor.
This behavior is inconvenient if the result bubbles are fullscreen so I needed to add some logic for when to scroll inside the bubble and when not.

@n-riesco Thanks for the help!

**Comment by n-riesco** [comment]
@lgeiger :+1: I should've read the code!

**Comment by hastebrot** [comment]
It would also be nice if the limit in `result-view.js` for the multiline result view could be configured in the settings view. `50` is currently hard-coded in `if (previousText === '' && text.length < 50`.

**Comment by lgeiger** [comment]
@hastebrot Thanks for chiming in. You are right, the hard-coded 50 characters are not great.

I think the best way would be to check if the inline output fits in the current view instead of making it user configurable. Though I'm not sure how this could be implemented.

Would you like to tackle this one? I'm happy to merge a PR 😄

**Comment by lgeiger** [comment]
At the moment I'm not using Hydrogen for my daily work. Could someone try out this PR and see how it feels?

**Comment by n-riesco** [comment]
Some feedback after taking this PR for a spin:

- the output box taking almost all the available width looks ugly for small outputs (see screenshot)

- action panel overlaps with large outputs (see screenshot)

- svg output is centered unlike all the other outputs (see screenshot)

- (this may be hard to fix, without breaking other functionality; e.g. autocomplete) but I find very distracting that the margin goes over the output boxes.

- I manage to get a `Uncaught TypeError: Cannot read property 'screenRow' of undefined`when scrolling up

- It'd be nice to have a way to distinguish output, stdout and stderr

screenshot from 2016-11-29 18-53-07

screenshot from 2016-11-29 18-54-27

screenshot from 2016-11-29 18-56-55

**Comment by lgeiger** [comment]
Thanks for taking a look!

> the output box taking almost all the available width looks ugly for small outputs

I did this on purpose to make the result bubbles feel more robust and visually consistent. I think the variable sizes can make the screen very cluttered. I wanted to make the multiline bubbles feel more like cells so we can support rich interactive output like plotly in the future.

> action panel overlaps with large outputs

Do you have a suggestion how we could fix it? The only option I see is to go back to a horizontal menu (which I tried to avoid).

> svg output is centered unlike all the other outputs 

I adopted the behavior of nteract here. We could align everything left if you like.

> I find very distracting that the margin goes over the output boxes.

Could you elaborate?

> I manage to get a Uncaught TypeError: Cannot read property 'screenRow' of undefinedwhen scrolling up

This is fixed in the latest Atom release: https://github.com/atom/atom/pull/13321

> It'd be nice to have a way to distinguish output, stdout and stderr

I think we would need to introduce multiple output containers to make this work. How does nteract and jupyter handle this?

**Comment by rgbkrk** [comment]
> > svg output is centered unlike all the other outputs
> I adopted the behavior of nteract here.

I feel like our centering is a bad idea for images inside of nteract/nteract. We should use the default image layout that comes from the browser

**Comment by n-riesco** [comment]
>  I wanted to make the multiline bubbles feel more like cells so we can support rich interactive output like plotly in the future.

I don't follow the reasoning there (i.e. I don't see what wouldn't allow the support of an interactive output). Having said this, please, take my comment just as one user's comment. If you don't see anything ugly, your opinion is more important.

> Do you have a suggestion how we could fix it?

padding?

> I adopted the behaviour of nteract here. We could align everything left if you like.

Again, take my comment as one user's comment.

> Could you elaborate?

margin


> How does nteract and jupyter handle this?

screenshot from 2016-11-29 19-58-32



---

One question: is `Hydrogen` planning to replicate `nteract`'s functionality?

**Comment by lgeiger** [comment]
I should have expressed my self a bit more carefully. Please don't see this PR as "my vision for Hydrogen". I originally only wanted to refactor the result view code to address #252.
My refactor completely broke the styling so I just started from scratch 😆.

> I don't follow the reasoning there

The main thing I don't like about the current output bubbles are their behavior with large output. For example multiple plots in one output bubble don't look good.
This is why I initially opted for full width bubbles. I definitely see your point with the full width output bubbles now (I just tried it with a external widescreen monitor 😄 ).

I'll try to revise the design next week to address your feedback.

> I find very distracting that the margin goes over the output boxes.

Oh the wrap-guide, yes that's distracting. But I don't know if we can do anything about that.

> It'd be nice to have a way to distinguish output, stdout and stderr

I think we should leave this for a later PR.

> One question: is Hydrogen planning to replicate nteract's functionality?

Oh no, I just used it for inspiration here. Maybe a bit too much 😄

**Comment by nikitakit** [comment]
I haven't had a chance to test the code yet, but I just want to quickly comment regarding the margin box. Have you tried limiting the width of the results box to the margin? This would help with scrolling, in that you could now use the area right of the margin to scroll the document instead of the results view.

It should also be the case that if you're actually following coding style and wrapping at the margin, you may want to open up multiple panes to make the most of your horizontal screen space. Any thoughts on this?

**Comment by lgeiger** [comment]
Thanks for all of your comments! 👍 

I'm going to close this PR because I tried to do too much at once.

That said, I learned a lot during the process and will split the improvements into smaller PRs.

**Pull request closed without merging**